---
title: "Statistical inference"
author: "Ruth Gómez Graciani"
date: "30 de octubre de 2017"
output: html_document
---

_Introduction_

_Specify the biological question you are asking._

Among the causes of the countless diseases that humans can suffer, genetic causes are ones of the most concerning because they are usually treatable but not fixable/curable and they can be transmitted to our offspring. However, genetic factors of diseases can vary from slight predispositions to totally letal mutations and rearrangements that lead to miscarriages even before the pregnancy is noticeable. 

This extremely deleterious mutations are hardly ever detected nor reported.


Assuming that there are differences in the essentiality of the genes (since the same mutation can be more or less severe deppending on the gene that it affects), we expect that the important genes are distributed randomly across the chromosomes, and the same can be applied to the less essential genes, as well as the mutations. Therefore, the source chromosome of a genetic disease is expected to depend only on the number of genes that can have a mutation in that chromosome. 


Why we are counting genes and not space
##

Are diseases evenly distributed across the chromosomes?

The null hipothesis regarding this question is that the number of diseases that a chromosome can be related to depends on the number of genes that ir contains, and the alternate hipothesis is that there are some disease-prone chromosomes. In statistical terms, these two hipothesis can be expressed as H0: all chromosomes are related to a number of diseases that depends exclusively on their number of genes and H1: some chromosmes have more reported diseases than the expected for their number of genes. 

<!-- _Put the question in the form of a biological null hypothesis and alternate hypothesis._ -->

<!-- H0: Todos los cromosomas tienen un numero aleatorio de enfermedades en función de su densidad génica y/o tamaño -->
<!-- H1: el factor 'cromosoma' también afectal número de enfermedades, y no solo la densidad génica o el tamaño -->
<!-- _Put the question in the form of a statistical null hypothesis and alternate hypothesis._ -->

<!-- H0: el número de enfermedades por gen está homogéneamente distribuido entre los cromosomas -->
<!-- h1: hay algunos cromosomas que tienen más enfermedades por gen de lo esperado -->

_Variables selection_

Number of diseases - discrete measure (separated by chromosomes)
Number of genes - discrete measure (Separated by chromosomes)
Cromosomes - nominal/ categories

_Design an experiment that controls or randomizes the confounding variables._

The population that will be studied
Enfermedades:
IDEAL FORMAT: Number of gene diseases, number of chromosomal diseases, for every chromosome (Mit and Unkown are excluded)
SOURCE: Disease database (OMIM) https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4383985/
WHY OMIM (PROS): It has updated and curated descriptions of human genes and phenotypes and the relationships between them,  genome coordinate searching and thesaurus-enhanced search term options. Source material for OMIM is the peer-reviewed biomedical literature. 
CONS in OMIM
METHOD
Correct by chunk size?
Correct by number of appearances in chromosome?

4148(clinical basis) cs_molecular_basis_exists:TRUE
4000 mutation (cs_molecular_basis:(mutation OR ( (mutant) OR (variation) OR (mutational) OR (mutant OR mutational OR variation) ) OR ( (variance OR variant) )))
148 not mutation (cs_molecular_basis_exists:TRUE) AND NOT (cs_molecular_basis:(mutation OR ( (mutant) OR (variation) OR (mutational) OR (mutant OR mutational OR variation) ) OR ( (variance OR variant) )))

all_entries

 * Arrange data*
 * Test for all diseases controlling by size and number of genes *
 * Test for mutational diseases controlling by number of genes *
 * Test for structural diseases controlling by size (and number of genes?) *
 * Statistical treatment  *
 
_Based on the number of variables, the kinds of variables, the expected fit to the parametric assumptions, and the hypothesis to be tested, choose the best statistical test to use._

Anovas or ancovas are not suitable because I only have one measure per group. A Kruskall-Wallis? Regression?


_If possible, do a power analysis to determine a good sample size for the experiment._

Power analysis

_Do the experiment._

* Arrange data*
```{r}

# 1. Exploratory data analysis (apply descriptive statistics and graphs for one o more 
#  variables) of non-trivial variables of the genome table. 
link <- 'http://mscbioinformatics.uab.cat/base/documents/bioinformaticsintranet1718/summary_human_genome2015_10_18D22_6.txt'

header<-read.table(link, nrows = 1, stringsAsFactors = F)

# there is a missing column name, as we can see in https://www.ncbi.nlm.nih.gov/genome/51?genome_assembly_id=214366
t<-read.table(link, skip = 1 , stringsAsFactors = F)
colnames(t)<-c(  as.character(header[1:4]), "INSDC", as.character(header[5:12]))

# Also, the last row has a typo
t[26,2:13]<-c(as.character(t[26,3:13]), as.character(t[27,1]))
t<-t[1:26,]

# The variables are transformed to numeric
t[,6:13]<-apply(t[,6:13], 2, function(x){
  x<-gsub(",","",x)
  as.numeric(x)
  })

# Add column OMIM diseases
disease<-read.table("/media/ubuntu/BIOPROJECT/BIOINFORMATICS/MODULE_2/01_Statistical_Inference/OMIM")
colnames(disease)<-c('Name', 'N.diseases')
t<- merge(t, disease, all= T)

```

_Examine the data to see if it meets the assumptions of the statistical test you chose (primarily normality and homoscedasticity for tests of measurement variables). If it doesn't, choose a more appropriate test._

* Descriptive analysis *

```{r}
# Univariate analysis in continuous variables (height and weight)

library(Weighted.Desc.Stat)

gensize<-sum(as.numeric(t$`Size_(Mb)`))
t$Rel_size = t$`Size_(Mb)`/gensize
sapply(t[6:13], summary)
apply(t[6:13], 2, function(x){
  w.mean(x[which(is.na(x)== FALSE)], t[which(is.na(x)== FALSE),"Rel_size"])
})

t$Genes_per_Mb<-t$Gene/t$`Size_(Mb)`
summary(t$Genes_per_Mb[1:24])
t$Pseudogene_per_Mb<-t$Pseudogene/t$`Size_(Mb)`
summary(t$Pseudogene_per_Mb[1:24])
t$ratio_gene_pseud=t$Genes_per_Mb/t$Pseudogene_per_Mb
summary(t$ratio_gene_pseud[1:24])

boxplot(t$ratio_gene_pseud)
library(ggplot2)
library(gridExtra)
require(reshape2)

ggplot(data = melt(t[1:24,15:18]), aes(x=factor(1), y=value)) + 
  geom_boxplot(aes(fill=variable))+
  facet_wrap(.~variable,  scales = 'free')+
  xlab("")+
  theme(axis.text.x=element_blank())
  

```

_Apply the statistical test you chose, and interpret the results._

```{r}

t$test <- t$N.diseases / t$Gene
anova(lm(N.diseases ~ Name+Gene, data = t))
t-test
sum(t$N.diseases, na.rm = T)
aov


kruskal.test(t$Pseudogene, as.factor(t$Name))

plot(t$`Size_(Mb)`, t$N.diseases)

```
* Kruskal walis *

_Communicate your results effectively, usually with a graph or table._

* Graph *

Exon bases table source:
http://seqanswers.com/forums/showthread.php?t=5298

Source NCBI:
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3116565/