---
title: "Statistical inference"
author: "Ruth GÃ³mez Graciani"
date: "30 de octubre de 2017"
output: 
  pdf_document: 
    latex_engine: xelatex
bibliography: molecular-population-genetics.bib

---

#Biological question
Among the causes of the countless diseases that humans can suffer, genetic causes are ones of the most concerning because they are usually treatable but not fixable/curable and they can be transmitted to our offspring. However, genetic factors of diseases can vary from slight predispositions to totally letal mutations and rearrangements that lead to miscarriages even before the pregnancy is noticeable. When regarded in the Ohta's nearly neutral theory of evolution context [@Casillas1003], it can be considered that these extremely deleterious mutations, which are hardly ever detected nor reported, are in the lower tail of the distribution, and that the mutations inducing slight predispositions are in the slightly deletrious zone. Therefore, what is known as 'genetic disease' must be around the center of the deleterious mutations distribution. 

From now on in this work, there will be considered as 'mutations' only those that are deleterious or slightly deleterious, and for simplification it will be assumed that they affect a gene (although it is known that some deleterious mutations can affect other regions like enhancers or epigenetic targets). 

The same mutation can be more or less severe depending on the gene that it affects, meaning that there are differences in the essentiality of the genes, and that any change can be more or less deleterious according to importance of the gene. Given two chromosomes of equal gene density, it would be more likely to have an extremely deleterious mutation in that with the larger number of important genes. That's why it is not expected to find chromosomes with a large proportion of important genes, but rather a random distribution of such genes.

In this context, the biological question is: are mutations causing diseases in fact evenly distributed across human chromosomes?


# Hypotheses and variables

The biological hypotheses (random distribution of important genes vs. chromosomes with a higher deleterious potential) can be tested using as an estimator of important genes the number of diseases that have been reported for each of the chromosomes. It is expected to be proportional to the number of genes in each chromosome, meaning that both measures will follow the same distribution. Therefore, the null statistical hypothesis is that the distribution of diseases across the chromosomes follows the same as the genes, and the alternative hypothesis is that some chromosomes have more (or less) diseases associated than the expected, in other words, that the distributions are different. 

# Experimental design
A priori, the chromosomes included in the study are autosomal and sex human chromosomes. Their general properties are published in [@Homo_sapiens]. An unbeatable source for the number of diseases for each chromosome is the OMIM database[@OMIM], because it has the information from the peer-reviewed biomedical literature updated and curated, resulting in reliable descriptions of human diseases and their most probably causing genes. It provides tools as genome coordinate searching and thesaurus-enhanced search term options. 

A possibly confounding variable is the length of the genes, because there is more space to suffer mutations. To assess that this is not affecting the analysis, it can be analised the correlation between the number of genes and the total exon length in bp, obtained from [@Exon]. If they are not significatively correlated, the study will have to be repeated for both measures (gene and exon length) and the results compared. 

Other confounding variables that can not be controlled are those related to the diseases; for instance rare diseases tend to be less studied that those that affect a wide portion of the population, or the studies are less reliable due to the sample size. In addition, not all kind of mutations, genes and chromosomes are equally easy to study. These factors can be partially randomized by having into account diseases with different amounts of knowledge associated. 


# The statistical test

There is available a list of number of observations (diseases) for each category in a nominal variable (chromosome), enough data to take a large sample (>100 observations, >5 observations per category), and it has to be tested the adequation of this data to an expected distribution, so in this case the best statistical test would be a Chi-square test of goodness-of-fit[@Chi]. It is also expected a fitness to the parametric assumptions but it's not mandatory in this test, so there's no need to confirm it.

As the data is available before the study, a post hoc power analysis can be performed after the data acquisition and cleaning to determine the best sample size. 

# Data preparation


```{r, echo = FALSE, message = FALSE, warning = FALSE}
  library(ggplot2)
  library(reshape2)
  library(gridExtra)
  library(pwr)

# BASIC TABLE

  link <- 'http://mscbioinformatics.uab.cat/base/documents/bioinformaticsintranet1718/summary_human_genome2015_10_18D22_6.txt'
  header<-read.table(link, nrows = 1, stringsAsFactors = F)
  
  # there is a missing column name, as we can see in https://www.ncbi.nlm.nih.gov/genome/51?genome_assembly_id=214366
  t<-read.table(link, skip = 1 , stringsAsFactors = F)
  colnames(t)<-c(  as.character(header[1:4]), "INSDC", as.character(header[5:12]))
  
  # Also, the last row has a typo
  t[26,2:13]<-c(as.character(t[26,3:13]), as.character(t[27,1]))
  t<-t[1:26,]
  
  # The variables are transformed to numeric
  t[,6:13]<-apply(t[,6:13], 2, function(x){
    x<-gsub(",","",x)
    as.numeric(x)
    })
  
  # Only the needed data is selected
  t<-t[c(1:24),c("Name","Size_(Mb)", "Gene")]
  t$Name <- factor(t$Name,  levels = c(1:22, 'X', 'Y'))


# DISEASES 

  t$Number_Diseases<-c(1638,1038,898,646,749,955,774,585,622,612,997,861,308,538,486,675,963,243,1035,427,172,374,848,54)

# EXON TOTAL LENGTH

  # Read table
  Basepairs<-read.table('Data/Exon_bases',skip=1, sep = ' ')
  colnames(Basepairs)<-c('Name', 'Total_Bases', 'Exon_Bases')
  Basepairs$Name<-sub('chr', '', Basepairs$Name)
  Basepairs$Name<-as.factor(Basepairs$Name)
  
  # Add to general table
  t<-merge(t, Basepairs[c(1:24),c(1,3)], by = 'Name', sort = FALSE)
  

```


All the data was incorporated to the R environment. After that, some descriptive analysis were performed:


```{r}

  sapply(t[2:5], summary)
  
```

In addition, the plot below summarizes all the numerical data with boxplots and barplots. They can be observed several outlier points corresponding to chromosome y for number of diseases and 1 for Genes and Exon Bases.





```{r, echo = FALSE, message = FALSE, warning=FALSE}
  # Summary plots
  box<-data.frame(t$Name)
  colnames(box)<-'Name'
  box$`Disease per gene`<-t$Number_Diseases/t$Gene
  box$Genes <-t$Gene
  box$`Exon Bases`<-t$Exon_Bases
  box<-melt(box)
  
  bx<-ggplot(data = box, aes(x=factor(1), y=value, fill = variable, group = variable)) +
    geom_boxplot()+
    facet_grid(variable ~.,  scales = 'free')+
    xlab("Box")+
    ylab("Count")+
    theme(axis.text.x=element_text(colour = 'white'),legend.position="none")
  
  br<-ggplot(data = box, aes(x=Name, y=value, fill = variable, group = variable)) +
   geom_bar(stat = 'identity')+
    facet_grid (variable ~ . , scales = 'free')+
    ylab("")+
    theme(axis.text.y=element_blank(), legend.position="none", axis.text.x = element_text(size = 6))

  
  grid.arrange(bx, br, ncol = 2)
```

Also, as Genes and Exon Bases are significatively correlated, the Chi square test can be performed using only Genes as a predictor (see Extra Plot 1). 

```{r}

  cor.test(t$Gene, t$Exon_Bases)
```

```{r, echo = FALSE}
 EP1<- ggplot(t, aes(x = Gene, y = Exon_Bases))+
    geom_point()+
    geom_smooth(method = 'lm')
  
```

Finally, the post hoc power test suggested that the a good sample size would be of around 800 diseases. This allows to have a power different from 1, which would be problematic because Chi-square tests are sensitive to the excess of data, giving false positives. 

``` {r}
  P0<-t$Gene/sum(t$Gene) 
  P1<-t$Number_Diseases/sum(t$Number_Diseases)

  effect.size <- ES.w1(P0, P1)
  
  pwr.chisq.test(w=effect.size,  N=NULL,  df=23,power=0.8,  sig.level=0.05) 

```

# Statistical test

First, a sample was acquired in order to have the right N for this test. It can be seen in that 800 individuals can be representative of the total (see Extra Plot 2). However, to mantain the low effect size, the original sample will be multiplied by 0.05 instead of sampling 800 individuals. 
```{r, echo = FALSE}

# RESAMPLING

  # Original data numbers
  Resample<-data.frame(t$Name, t$Number_Diseases, rep('Original'), t$Number_Diseases/sum(t$Number_Diseases))
  colnames(Resample)<-c('Name', 'Diseases', 'Group', 'Proportion')
  
  # Simulation of the raw data
  population<-unlist(apply(t, 1, function(x){
    rep(x[1],x[4] )
  }))

  # Sample from the raw data
  Sample1<-data.frame(table(sample(population, 800, replace = FALSE)), rep('Sample'))
  # Sample<-data.frame(t$Name, t$Number_Diseases*0.05, rep('Sample'))
  colnames(Sample1)<-c('Name', 'Diseases', 'Group')
  Sample1$Proportion<-Sample1$Diseases/sum(Sample1$Diseases)
  Resample1<-rbind(Resample, Sample1)
  
   # Sample<-data.frame(table(sample(population, 800, replace = FALSE)), rep('Sample'))
  Sample<-data.frame(t$Name, t$Number_Diseases*0.05, rep('Sample'))
  colnames(Sample)<-c('Name', 'Diseases', 'Group')
  Sample$Proportion<-Sample$Diseases/sum(Sample$Diseases)
  Resample<-rbind(Resample, Sample)
 
```

```{r, echo = FALSE}
  EP2<-ggplot(Resample1, aes(x = Name, y = Proportion, color=Group, group = Group, line = Group))+geom_line()+geom_point()

    # TEST Y has a very low number of observations, so it's deleted
  # 
  Observed<-Sample$Diseases
  Expected<-t$Gene
  Expected_Frq<-Expected/(sum(Expected))
  
```

After this,the new was also calculated to adjust the significance level.

```{r}

  chisq.test(x=Observed, p =Expected_Frq)
  
# NEW POWER
 
  P0<-Expected_Frq
  P1<-Observed/sum(Observed)

  effect.size <- ES.w1(P0, P1)
  
  pwr.chisq.test( w=effect.size,N=sum(Observed), df = 23, power= 0.8,   sig.level = NULL)    
  
```

Afther these analysis, we can't reject the null hypothesis, meaning that both observed and expected distributions are the same (See Estra Plot 3) and taht there's no apparent presence of  a chromosome that causes more diseases than expected.

# Extra plots

## Extra Plot 1
```{r, echo = FALSE, fig.height=3}
EP1
```

## Extra Plot 2

```{r, echo = FALSE, fig.height=2}
EP2
```

## Extra Plot 3

```{r, echo = FALSE, message= FALSE, warning=FALSE, fig.height=2}


library(reshape2)
# 
t$Gene_prop<-t$Gene/sum(t$Gene)
t$Dis_prop<-t$Number_Diseases/sum(t$Number_Diseases)
Pl<-melt(t[, c("Name", "Gene_prop", "Dis_prop")])


ggplot(data=Pl, aes(x=Name, y=value, group=variable, color = variable, linetype=variable)) +
  geom_line()+
  geom_point( size=1)
```

# Bibliography