
* Population genetics expectations---
title: "Statistical inference"
author: "Ruth GÃ³mez Graciani"
date: "30 de octubre de 2017"
output: html_document
---

_Abstract_

_Specify the biological question you are asking._

Among the causes of the countless diseases that humans can suffer, genetic causes are ones of the most concerning because they are usually treatable but not fixable/curable and they can be transmitted to our offspring. However, genetic factors of diseases can vary from slight predispositions to totally letal mutations and rearrangements that lead to miscarriages even before the pregnancy is noticeable. When regarded in the Ohta's nearly neutral theory of evolution context, we can consider that these extremely deleterious mutations, which are hardly ever detected nor reported, are in the lower tail of the distribution, and that the mutations inducing slight predispositions are in the slightly deletrious zone. Therefore, what we know as 'genetic disease' must be around the center of the deleterious mutations distribution. 

From now on in this work, if there's no further [clarification], we'll consider as 'mutations' only those that are deleterious or slightly deleterious, and for simplification we assume that they affect a gene (although it is known that some deleterious mutations can affect other regions like enhancers or epigenetic targets). 

The same mutation can be more or less severe depending on the gene that it affects, so we can assume that there are differences in the essentiality of the genes, and that any change can be more or less deleterious according to importance of the gene. Given two chromosomes of equal gene density, it would be more likely to have an extremely deleterious mutation in that with the larger number of important genes. That's why we don't expect chromosomes with a large proportion of important genes, but rather a random distribution of such genes.

In this context, the biological question is: are mutations causing diseases in fact evenly distributed across human chromosomes?

```{r}
<!-- Among the causes of the countless diseases that humans can suffer, genetic causes are ones of the most concerning because they are usually treatable but not fixable/curable and they can be transmitted to our offspring. However, genetic factors of diseases can vary from slight predispositions to totally letal mutations and rearrangements that lead to miscarriages even before the pregnancy is noticeable.  -->

<!-- This extremely deleterious mutations are hardly ever detected nor reported. -->

<!-- Assuming that there are differences in the essentiality of the genes (since the same mutation can be more or less severe deppending on the gene that it affects), we expect that the important genes are distributed randomly across the chromosomes, and the same can be applied to the less essential genes, as well as the mutations. Therefore, the source chromosome of a genetic disease is expected to depend only on the number of genes that can have a mutation in that chromosome.  -->


<!-- Why we are counting genes and not space -->
<!-- ## -->

<!-- Are diseases evenly distributed across the chromosomes? -->

<!--  -->
```
 
 _Put the question in the form of a null hypothesis and alternate hypothesis._ _Variables selection_

The biological hypotheses (random distribution of important genes vs. chromosomes with a higher deleterious potential) can be tested using as an estimator of important genes the number of diseases that have been reported for each of the chromosomes. It is expected to be proportional to the number of genes in each chromosome, meaning that both measures will follow the same distribution. Therefore, our null statistical hypothesis is that the distribution of diseases across the chromosomes follows the same as the genes, and the alternate hypothesis is that some chromosomes have more (or less) diseases associated than the expected, thus the distributions are different. 

_Design an experiment that controls or randomizes the confounding variables._

A priori, the chromosomes included in the study are autosomal chromosomes and sex human chromosomes.  Their general properties are published in [source]. An unbeatable source for the number of diseases for each chromosome is the OMIM database, because it has the information from the peer-reviewed biomedical literature updated and curated, resulting in reliable descriptions of human diseases and their most probably causing genes. It provides tools as genome coordinate searching and thesaurus-enhanced search term options. 

A possibly confounding variable is the length of the genes, because there is more space to suffer mutations. To assess that this is not affecting the analysis, it will be analised the correlation between the number of genes and the total exon length in bp. If they are not significatively correlated, the study will have to be repeated for both measures (gene and exon length) and the results compared. 

Other confounding variables that can not be controlled are those related to the diseases; for instance rare diseases tend to be less studied that those that affect a wide portion of the population, or the studies are less reliable due to the sample size. In addition, not all kind of mutations, genes and chromosomes are equally easy to study. These factors can be partially randomized by having into account diseases with different amounts of knowledge associated. 

```{r}

<!-- The experiment is the process of obtainig measurement data, in this case, from databases. The general information about the chromosomes is published in [source]. In addition, the OMIM database is a suitable source of the number of diseases per chromosome, as it has updated and curated descriptions of human genes and phenotypes and the relationships between them, and genome coordinate searching and thesaurus-enhanced search term options. Source material for OMIM is the peer-reviewed biomedical literature.  -->

<!-- I have been able to identify two main cofounding variables: chromosome size/gene density, chromosome ease of study and mutation location ease of study, disease ease of study.  -->

<!-- The chromosome size will be contorlled by using the distribution of the genes across the chromosome as an estimator of the distribution that the number of diseases should have. The ease of study of each chromosome is randomized by having into account all the possible entries intead of only the most complete ones. I can't correct the mutation location, there are some disease that are intergenic, etc, but most of them will have effect in some gene, so it's ok. Also rare diseases are not so widely studied and common diseases not, but it's ok -->


<!-- 4148(clinical basis) cs_molecular_basis_exists:TRUE -->
<!-- 4000 mutation (cs_molecular_basis:(mutation OR ( (mutant) OR (variation) OR (mutational) OR (mutant OR mutational OR variation) ) OR ( (variance OR variant) ))) -->
<!-- 148 not mutation (cs_molecular_basis_exists:TRUE) AND NOT (cs_molecular_basis:(mutation OR ( (mutant) OR (variation) OR (mutational) OR (mutant OR mutational OR variation) ) OR ( (variance OR variant) ))) -->
<!-- c(1,) -->

<!-- (tx_molecular_genetics_exists:TRUE) AND (cs_molecular_basis_exists:TRUE) AND  (cs_molecular_basis:mutation) AND (cs_inheritance_exists:TRUE) AND NOT (cs_inheritance:Mitochondrial) -->

<!--  chromosome_group:A AND tx_molecular_genetics_exists:TRUE AND (cs_molecular_basis_exists:TRUE) AND  (cs_molecular_basis:mutation) AND (cs_inheritance_exists:TRUE) AND NOT (cs_inheritance:Mitochondrial) -->

<!-- all_entries -->


<!--  * Arrange data* -->
<!--  * Test for all diseases controlling by size and number of genes * -->
<!--  * Test for mutational diseases controlling by number of genes * -->
<!--  * Test for structural diseases controlling by size (and number of genes?) * -->
<!--  * Statistical treatment  * -->
```


_Based on the number of variables, the kinds of variables, the expected fit to the parametric assumptions, and the hypothesis to be tested, choose the best statistical test to use._

We will have a list of number of observations (diseases) for each category in a nominal variable (chromosome), enough data to take a large sample (>100 observations, >5 observations per category), and we want to test the adequation of our data to an expected distribution, so in this case the best statistical test would be a Chi-square test of goodness-of-fit. http://www.biostathandbook.com/chigof.html. It is also expected a fitness to the parametric assumptions but it's not mandatory, so it won't be tested. 

_If possible, do a power analysis to determine a good sample size for the experiment._

As our data is available before the study, a post hoc power analysis will be performed after the data acquisition and cleaning to determine the best sample size. 

_Data acquisition_


```{r}

# BASIC TABLE: 

link <- 'http://mscbioinformatics.uab.cat/base/documents/bioinformaticsintranet1718/summary_human_genome2015_10_18D22_6.txt'
header<-read.table(link, nrows = 1, stringsAsFactors = F)

# there is a missing column name, as we can see in https://www.ncbi.nlm.nih.gov/genome/51?genome_assembly_id=214366
t<-read.table(link, skip = 1 , stringsAsFactors = F)
colnames(t)<-c(  as.character(header[1:4]), "INSDC", as.character(header[5:12]))

# Also, the last row has a typo
t[26,2:13]<-c(as.character(t[26,3:13]), as.character(t[27,1]))
t<-t[1:26,]

# The variables are transformed to numeric
t[,6:13]<-apply(t[,6:13], 2, function(x){
  x<-gsub(",","",x)
  as.numeric(x)
  })

# Only the needed data is selected
t<-t[,]

# DISEASES 

t$Number_Diseases<-c(1638,1038,898,646,749,955,774,585,622,612,997,861,308,538,486,675,963,243,1035,427,172,374,848,54, NA, NA)

# EXON
Basepairs<-read.table('Data/Exon_bases',skip=1, sep = ' ')
colnames(Basepairs)<-c('Name', 'Total_Bases', 'Exon_Bases')
Basepairs$Name<-sub('chr', '', Basepairs$Name)
t<-merge(t, Basepairs)
# Add column OMIM diseases
# disease<-read.table("/media/ubuntu/BIOPROJECT/BIOINFORMATICS/MODULE_2/01_Statistical_Inference/OMIM")
# colnames(disease)<-c('Name', 'N.diseases')
# t<- merge(t, disease, all= T)



```

WHY OMIM (PROS
_Examine the data to see if it meets the assumptions of the statistical test you chose (primarily normality and homoscedasticity for tests of measurement variables). If it doesn't, choose a more appropriate test._

* Descriptive analysis * - search outliers
* Normality
* Homoscedasticity *
* Adequation of the Chisquare test*
* Resampling *
* Real power *

```{r}
# Univariate analysis in continuous variables (height and weight)

# library(Weighted.Desc.Stat)
# 
# gensize<-sum(as.numeric(t$`Size_(Mb)`))
# t$Rel_size = t$`Size_(Mb)`/gensize
# sapply(t[6:13], summary)
# apply(t[6:13], 2, function(x){
#   w.mean(x[which(is.na(x)== FALSE)], t[which(is.na(x)== FALSE),"Rel_size"])
# })
# 
# t$Genes_per_Mb<-t$Gene/t$`Size_(Mb)`
# summary(t$Genes_per_Mb[1:24])
# t$Pseudogene_per_Mb<-t$Pseudogene/t$`Size_(Mb)`
# summary(t$Pseudogene_per_Mb[1:24])
# t$ratio_gene_pseud=t$Genes_per_Mb/t$Pseudogene_per_Mb
# summary(t$ratio_gene_pseud[1:24])
# 
# boxplot(t$ratio_gene_pseud)
# library(ggplot2)
# library(gridExtra)
# require(reshape2)
# 
# ggplot(data = melt(t[1:24,15:18]), aes(x=factor(1), y=value)) + 
#   geom_boxplot(aes(fill=variable))+
#   facet_wrap(.~variable,  scales = 'free')+
#   xlab("")+
#   theme(axis.text.x=element_blank())
  

```

_Apply the statistical test you chose, and interpret the results._

* Application of the statistical test*
```{r}
# Expected_total<-sum(t$Number_Diseases)/sum(t$Gene)
# t$Expected_probs<-(t$Gene*Expected_total)/sum(t$Number_Diseases)
# 
# chi <-sum(((t$Number_Diseases-(t$Gene*Expected_total))^2)/(t$Gene*Expected_total))
# df = 22
# 1-pchisq(chi, 22)
# 
# 
# chisq.test(t$Number_Diseases, t$Expected_probs, simulate.p.value = T)
# chisq.test(t$Number_Diseases, t$Expected_probs)

# ###
# 
# library(pwr)
# 
# # P0<-t$Expected_probs
# P0<-t$Gene/sum(t$Gene)
# P1<- t$Number_Diseases/sum(t$Number_Diseases)
# 
# effect.size <- ES.w1(P0, P1)  
# 
# degrees <- 529
# 
# pwr.chisq.test(
#                w=effect.size, 
#                N=sum(t$Number_Diseases),            # Total number of observations
#                df=529, 
#                power=NULL,        # 1 minus Type II probability
#                sig.level=0.05)    # Type I probability
# 
# cor.test(t$Gene, t$Number_Diseases)
# plot(t$Gene, t$Number_Diseases)


# APPLY EXAMPLE
library(ggplot2)
Analysis<-data.frame(t$Name, t$Number_Diseases, rep('Original'))
Analysis1<-data.frame(t$Name, t$Number_Diseases*0.1, rep('01'))
Analysis5<-data.frame(t$Name, t$Number_Diseases*0.5, rep('05'))
colnames(Analysis)<-colnames(Analysis1)<-colnames(Analysis5)<-c('Name', 'Diseases', 'Group')
Analysis<-rbind(Analysis, Analysis1, Analysis5)

# p<-ggplot(Analysis, aes(x = Name, y = Diseases, color=Group))
# p+geom_line(group=1)+facet_grid(Group~., scales = 'free')
chisq.test(t$Number_Diseases*0.01, p=(t$Gene/t$`Size_(Mb)`)/(sum(t$Gene/t$`Size_(Mb)`)))

population<-unlist(apply(t, 1, function(x){
  rep(x[1],x[14] )
}))
poptab<-data.frame(table(sample(population, 1000, replace = FALSE)))
poptab2<-data.frame(table(sample(population, 1000, replace = FALSE)))
poptab4<-data.frame(table(sample(population, 1000, replace = FALSE)))
poptab1<-data.frame(table(sample(population, 1000, replace = FALSE)))
poptab3<-data.frame(table(sample(population, 1000, replace = FALSE)))

T<-cbind(poptab, poptab1, poptab2, poptab3, poptab4)
T<-T[,c(1,2,4,6,8,10)]
T$mean<-apply(T,1,FUN=function(x){mean(as.numeric(x[2:6]))})


Analysis_sam<-data.frame(T[,c("Var1", "mean")], rep('sample'))
genes<-data.frame(t$Name, t$Gene/t$`Size_(Mb)`, rep('Genes'))
colnames(Analysis_sam)<-colnames(genes)<-c('Name', 'Diseases', 'Group')
Analysis<-rbind(Analysis, Analysis_sam, genes)


p<-ggplot(Analysis, aes(x = Name, y = Diseases, color=Group))
p+geom_line(group=1)+facet_grid(Group~., scales = 'free')


library(pwr)

 
effect.size <- ES.w1((t$Gene/t$`Size_(Mb)`)/(sum(t$Gene/t$`Size_(Mb)`)), t$Number_Diseases/sum(t$Number_Diseases) )
# effect.size <-0.1
degrees <- 22
# 
pwr.chisq.test(
               w=effect.size,
               N=sum(t$Number_Diseases*0.01),            # Total number of observations
               df=degrees,
               power=0.9,        # 1 minus Type II probability
               sig.level=NULL)    # Type I probability
# 
# cor.test(t$Gene, t$Number_Diseases)
# plot(t$Gene, t$Number_Diseases)

library(reshape2)
# 
t$Gene_prop<-t$Gene/sum(t$Gene)
t$Dis_prop<-t$Number_Diseases/sum(t$Number_Diseases)
Pl<-melt(t[, c(1,17,18)])

ggplot(Pl, aes(x=Name, y = value, fill=variable))+geom_bar(position = 'dodge', stat = 'identity')
qqplot(t$Gene_prop, t$Dis_prop)

boxplot(t$Gene_prop/t$Dis_prop)
t$Ratio<-t$Gene_prop/t$Dis_prop
```
* According to the expectations*
 
_Communicate your results effectively, usually with a graph or table._

* Graphs and tables *

_Sources_

Exon bases table source: HG19
http://seqanswers.com/forums/showthread.php?t=5298

Source NCBI:
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3116565/

BARBADILLA - foto
https://www.ncbi.nlm.nih.gov/pubmed/28270526